### ISSUE TITLE
Search with specific index type is 3 time slower than no index type specified

### ISSUE THREAD

**Issue by sunchanras** [elastic/elasticsearch#14179]
Version 2.0rc1
Hardware; Core I5 ,8GB of Ram and SSD
________________________________________________________________________________
Here I have index:user and type:test with 40m documents index.
query1: /user/test/_search?q=title:how to install window10 
take around 2300ms to 2500ms

query2: /user/_search?q=title:how to install window10 
take around 700ms to 800ms
________________________________________________________________________________
This is normal behavior of elasticsearch? or bug on 2.0 version?

**Comment by clintongormley** [comment]
@sunchanras your first query also has to do a filter on `_type:test`.  Did you run this several times before comparing times? The filter should end up being cached (and fast).

**Comment by sunchanras** [comment]
@clintongormley  Yes, I run it several times.I have just tested with version 1.7 with 200 millions docs, the same result. For single index and single type;search with specific type is much slower than not specified type. I did not do filter on _type:test, I will test on that and will update result.

**Comment by clintongormley** [comment]
@jpountz any ideas here?

**Comment by sunchanras** [comment]
I did filter query with type "test"
{
  "query": {
    "filtered": {
      "query": {
        "term": {
          "title": "title:how to install window10"
        }
      },
      "filter": {
        "type": {
          "value": "test"
        }
      }
    }
  }
}
Same result.
200million docs, single index "user", single type "test"

**Comment by jpountz** [comment]
@sunchanras can you let us know about the following:
 - how many types do you have?
 - how many hits do both queries match?
 - what does the [validate API}(https://www.elastic.co/guide/en/elasticsearch/guide/current/_validating_queries.html) return when `explain` is turned on? (once with the type and once without please)

**Comment by sunchanras** [comment]
@jpountz 
 *  only 1 type "test"
 *  hit with specific type and non-specific type 6708012 
 *  I did not turn 'explain' on ( I will try again and will update later)
 *  mapping is default dynamic mapping

**Comment by sunchanras** [comment]
@jpountz 
query body:
```javascript
{
  "explain": true,
  "query": {
    "filtered": {
      "query": {
        "match": {
          "title": "how to install cassandra"
        }
      },
      "filter": {
        "type": {
          "value": "test"
        }
      }
    }
  }
}
```
response:
```javascript
{
  "valid": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "failed": 0
  }
}
```
query validation false but no explanation.
Thank

**Comment by jpountz** [comment]
Sorry I was not clear, what I would like to see is the output of:

```
GET /user/test/_validate/query?q=title:how to install window10&explain
```
and
```
GET /user/_validate/query?q=title:how to install window10&explain
```

It looks like the optimization for disjunction we have when they are used at the top-level (usually referred to as BS1) makes things slower than the regular scorer (usually referred to as BS2), which is quite surprising given how BS1 is usually much faster than BS2.

**Comment by sunchanras** [comment]
```
 /user/test/_validate/query?q=title:how to install window10&explain

{
  "valid": true,
  "_shards": {
    "total": 1,
    "successful": 1,
    "failed": 0
  },
  "explanations": [
    {
      "index": "user",
      "valid": true,
      "explanation": "+(title:how _all:to _all:install _all:window10) #QueryWrapperFilter(+ConstantScore(_type:test))"
    }
  ]
}
```

```
/user/_validate/query?q=title:how to install window10&explain


{
  "valid": true,
  "_shards": {
    "total": 1,
    "successful": 1,
    "failed": 0
  },
  "explanations": [
    {
      "index": "user",
      "valid": true,
      "explanation": "title:how _all:to _all:install _all:window10"
    }
  ]
}
```
Thank

**Comment by sunchanras** [comment]
@jpountz 
Here the captured image from Kopf
1.with specific type 
![e1](https://cloud.githubusercontent.com/assets/6205717/10611856/4af2f316-7778-11e5-998e-7955aa041f0e.jpg)

2. without specific type
![e2](https://cloud.githubusercontent.com/assets/6205717/10611860/504bdc9c-7778-11e5-90de-4c5a5d3bf33c.jpg)

Thanks

**Comment by jpountz** [comment]
@sunchanras Any chance that you can capture [hot threads](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-nodes-hot-threads.html) while the slow query (the index-level one) is running? This could help figure out the bottleneck. Please put the content in a gist and link it here.

**Comment by jpountz** [comment]
Also out of curiosity, how many [segments](https://www.elastic.co/guide/en/elasticsearch/reference/1.7/indices-segments.html) does your shard have?

**Comment by sunchanras** [comment]
@jpountz 
For segments
```
"num_committed_segments": 28,
"num_search_segments": 28,
```
Thank

**Comment by sunchanras** [comment]
@jpountz 
For hot thread
https://gist.github.com/sunchanras/e0d2107b62aa884ce561
Sorry I am not sure what "the index-level one" is.
Thank

**Comment by jpountz** [comment]
Hmm, the hot threads are empty, are you sure you generated them while the query was running? By index-level, I meant the query where the type is not specified.

**Comment by jpountz** [comment]
It might help to run the query in a loop and capture the hot threads several times to get a better idea of what is happening.

**Comment by sunchanras** [comment]
OK, thank. I will run with loop and will update again.

**Comment by sunchanras** [comment]
@jpountz I have just updated the gist, and you view the gist again, please?
https://gist.github.com/sunchanras/e0d2107b62aa884ce561
Thank.

**Comment by sunchanras** [comment]
@jpountz 
Here hot-thread for specific type query
```
 /user/test/_search?q=title:how to install window10&explain
```

https://gist.github.com/sunchanras/7c2ee8b994ed4981f5d7

Thank

**Comment by jpountz** [comment]
Thank you, this is very helpful. So it looks like in your case BS2 is actually faster than BS1, which is quite unusual. Can you please run the following queries to see how many documents contain each individual term?

```
GET /user/_search
{
  "size": 0,
  "query": {
    "term": {
      "title": "how"
    }
  }
}
```

```
GET /user/_search
{
  "size": 0,
  "query": {
    "term": {
      "_all": "to"
    }
  }
}
```

```
GET /user/_search
{
  "size": 0,
  "query": {
    "term": {
      "_all": "install"
    }
  }
}
```

```
GET /user/_search
{
  "size": 0,
  "query": {
    "term": {
      "_all": "windows10"
    }
  }
}
```

**Comment by sunchanras** [comment]
hit:0 doc
```
GET /user/_search
{
  "size": 0,
  "query": {
    "term": {
      "title": "how"
    }
  }
}
```

hit:30758725 docs
```
GET /user/_search
{
  "size": 0,
  "query": {
    "term": {
      "_all": "to"
    }
  }
}
```

hit:0 docs
```
GET /user/_search
{
  "size": 0,
  "query": {
    "term": {
      "_all": "install"
    }
  }
}
```

hit:0 docs
```
GET /user/_search
{
  "size": 0,
  "query": {
    "term": {
      "_all": "windows10"
    }
  }
}

```

Thank

**Comment by sunchanras** [comment]
@jpountz 
I created a function to generate arbitrary title from wiki docs randomly to index.
```
String string1=[word1,word2,word3...]
String string2=[word1,word2,word3...]
String string3=[word1,word2,word3...]
String string4=[word1,word2,word3...]
String string5=[word1,word2,word3...]
String string6=[word1,word2,word3...]
String title=string1+string2+string3+string1+string5+string6;
```
Thank

**Comment by sunchanras** [comment]
@jpountz 
You can setup a test with single index and single type, 40 million docs.
You will see big different between query with or without specific type.

Thank.

**Comment by sunchanras** [comment]
@jpountz 
I am new to ES, thought. I just moved from SOLR.
ES is great in nearly every aspect for distributed search.

**Comment by jpountz** [comment]
Thank you for helping me understand what is happening here. When filtering by a type, Lucene manages to understand that it really only has to run a single term query since there is a single clause that matches documents. However when the query is not filtered, it still tries to apply the logic that we use for disjunctions, which adds unnecessary overhead. I opened an issue against Lucene: https://issues.apache.org/jira/browse/LUCENE-6850 and will close this one.

**Comment by sunchanras** [comment]
@jpountz 
Regarding to https://issues.apache.org/jira/browse/LUCENE-6850 , was it fixed on ES 2.0 ? So I can set up test again.
Thank.

**Comment by nik9000** [comment]
@sunchanras that looks like its in Lucene 6.0, 5.4. Elasticsearch [2.0](https://github.com/elastic/elasticsearch/blob/2.0/pom.xml) uses Lucene 5.2.1. Elasticsearch [2.1](https://github.com/elastic/elasticsearch/blob/2.1/pom.xml) will Lucene use 5.3.0. Elasticsearch [2.2](https://github.com/elastic/elasticsearch/blob/2.x/pom.xml) and [3.0](https://github.com/elastic/elasticsearch/blob/master/gradle.properties) plans to use Lucene 5.4.0.
